### PR TITLE
Add shift and range expression parsing

### LINKED ISSUE #22953: Add shift expression parsing
**Description:**
Refer: https://ballerina.io/spec/lang/2020R1/#shift-expr

Note: this will have an impact on type-cast-expr. e.g.: `< int > a < < int > b`

**Suggested Labels:**
Component/Parser, Type/NewFeature

### LINKED ISSUE #23266: Add range expression parsing
**Description:**
Refer: https://ballerina.io/ballerina-spec/spec.html#range-expr

**Suggested Labels:**
Component/Parser, Type/NewFeature

**Suggested Assignees**
@lochana-chathura

### LINKED ISSUE #23287: Parser crashes with type-cast-expr recovery
**Description:**

Sample: 
```java
function foo(){
    a = <int> > a;
}
```

Error: 
```
Exception in thread "main" java.lang.IllegalStateException: type-cast-expr
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExpressionRhs(BallerinaParserErrorHandler.java:1431)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:659)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:188)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:375)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:316)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1263)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:188)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:368)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:316)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1263)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:188)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:368)
```

Also operator precedence is wrong in `type-cast-expr` parsing.

Sample:
```
function foo(){
    a = <int> a + <int> b;
}
```

**Suggested Labels:**
Component/Parser, Priority/Blocker, Type/Bug

**Suggested Assignees:**
@lochana-chathura

### PR DESCRIPTION
**Pull request by lochana-chathura**
Fixes #22953  
Fixes #23266  
Fixes #23287 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -3684,6 +3700,17 @@ private STNode parseExpressionRhs(SyntaxKind tokenKind, OperatorPrecedence curre
...
+        // TODO: Add a good comment
+        if (tokenKind == SyntaxKind.LT_TOKEN && peek(2).kind == SyntaxKind.LT_TOKEN) {
+            tokenKind = SyntaxKind.DOUBLE_LT_TOKEN;
+        } else if (tokenKind == SyntaxKind.GT_TOKEN && peek(2).kind == SyntaxKind.GT_TOKEN) {
+            if (peek(3).kind == SyntaxKind.GT_TOKEN) {
+                tokenKind = SyntaxKind.TRIPPLE_GT_TOKEN;
+            } else {
+                tokenKind = SyntaxKind.DOUBLE_GT_TOKEN;
+            }
+        }
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3712]: In here we have to check whether next tokens belong to a `type-cast-expr`. It seems it is not possible unless we process up to the `>`. i.e. next token being an `@` or `isTypeStartingToken` does not always imply a `type-cast-expr` start.

Even if we processed up to `>`, still there's ambiguity. 
e.g.  `a <<b> c`  could mean either `a << (b > c)` or `a < (<b> c)`

So feel like it is better not  allow `a <<int> b;` to be parsed as a `type-cast-expr`.
Only  `a < <int> b;` is allowed. That way we don't need nodes for `<<`, `>>`, `>>>` at parser level. WDT?

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:3712]: According to the spec, type-cast has higher-precedence over shift/compare expressions. So `a <<b> c` is correct to be parsed as `a <(<b> c)`. If the user wants the other behaviour, he can use parenthesis.

But yea, like you've said, we have to go and parse the component after `<` to determine the operator. Let me have a look and get back.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -3684,6 +3700,17 @@ private STNode parseExpressionRhs(SyntaxKind tokenKind, OperatorPrecedence curre
             }
         }
 
+        // TODO: Add a good comment
+        if (tokenKind == SyntaxKind.LT_TOKEN && peek(2).kind == SyntaxKind.LT_TOKEN) {
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: We hv operator precedence check in between which decides whether to return from method or continue. We have to process >>, >>> before that check.

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/TypeCastExpressionTest.java [hunk 3]
```diff
@@ -41,13 +41,18 @@ public void testTypeCast() {
...
+    }
+
     // Recovery tests
 
     @Test
     public void testTypeCastWithMissingExpr() {
-        test("<int>", "type-cast-expr/type_cast_expr_assert_08.json");
-        test("<@foo{}>", "type-cast-expr/type_cast_expr_assert_09.json");
-        test("<@foo{} int>", "type-cast-expr/type_cast_expr_assert_10.json");
+        test("<int>;", "type-cast-expr/type_cast_expr_assert_08.json");
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/TypeCastExpressionTest.java:53]: In type-cast-expr recovery ctx ends after` >` token is passed. But I had accidentally pased expression also under type-cast ctx which is now fixed in this PR. Refer https://github.com/ballerina-platform/ballerina-lang/issues/23287

Without the `;` above test cases passed with that mistake.

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/types/ErrorTypeTest.java [hunk 4]
```diff
@@ -73,7 +73,7 @@ public void testInValidErrorTypeMissingErrorTypeParams() {
 
     @Test
     public void testInValidErrorTypeExtraGt() {
-        test("error<NO_MATCHING_OBJECT>> a;", "error-type/error_type_assert_02.json");
+        test("error<NO_MATCHING_OBJECT> > a;", "error-type/error_type_assert_10.json");
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/types/ErrorTypeTest.java]: yeah. This was before I moved >>, >>> parsing from lexer. It is now fixed

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/TypeCastExpressionTest.java [hunk 5]
```diff
@@ -41,13 +41,18 @@ public void testTypeCast() {
...
+        test("<@foo{}>;", "type-cast-expr/type_cast_expr_assert_09.json");
+        test("<@foo{} int>;", "type-cast-expr/type_cast_expr_assert_10.json");
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/TypeCastExpressionTest.java:55]: few new test cases added

File: compiler/ballerina-parser/src/test/resources/expressions/func-call/func_call_assert_11.json [hunk 6]
```diff
@@ -58,52 +58,47 @@
...
-                                                    "isMissing": true
-                                                },
-                                                {
-                                                    "kind": "ELLIPSIS_TOKEN"
-                                                },
-                                                {
-                                                    "kind": "IDENTIFIER_TOKEN",
-                                                    "value": "subjects"
-                                                }
-                                            ]
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/test/resources/expressions/func-call/func_call_assert_11.json:106]: With `...` rage operator var-arg in  `foo( bar(5, a, age  baz() ...subjects), 5, a, age = 18, ...subjects)` is now recovered as range expression.

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/types/ErrorTypeTest.java [hunk 7]
```diff
@@ -73,7 +73,7 @@ public void testInValidErrorTypeMissingErrorTypeParams() {
 
     @Test
     public void testInValidErrorTypeExtraGt() {
-        test("error<NO_MATCHING_OBJECT>> a;", "error-type/error_type_assert_02.json");
+        test("error<NO_MATCHING_OBJECT>> a;", "error-type/error_type_assert_10.json");
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/types/ErrorTypeTest.java:76]: `error_type_assert_02.json` is also used in another test case as well. So I added `10.json` here which is identical to `02.json`

**Comment by lochana-chathura** [comment]
@SupunS Conflicts are resolved!

File: compiler/ballerina-parser/src/test/resources/expressions/func-call/func_call_assert_11.json (hunk 6 above)
Comment (SupunS) [reply compiler/ballerina-parser/src/test/resources/expressions/func-call/func_call_assert_11.json:106]: should be fine

**Pull request merged**